actionners/kubernetes/drain/drain.go

```diff
@@ -0,0 +1,188 @@
+	"github.com/go-playground/validator/v10"
+	corev1 "k8s.io/api/core/v1"
+	policyv1 "k8s.io/api/policy/v1"
+	policyv1beta1 "k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"regexp"
+	"strings"
+)
+
+const (
+	validatorName = "is_absolut_or_percent"
+	// EvictionKind represents the kind of evictions object
+	EvictionKind = "Eviction"
+	// EvictionSubresource represents the kind of evictions object as pod's subresource
+	EvictionSubresource = "pods/eviction"
+)
+
+type Config struct {
+	MinHealthyReplicas string `mapstructure:"min_healthy_replicas" validate:"omitempty,is_absolut_or_percent"`
+	IgnoreDaemonsets   bool   `mapstructure:"ignore_daemonsets" validate:"omitempty"`
+	IgnoreStatefulSets bool   `mapstructure:"ignore_statefulsets" validate:"omitempty"`
+	GracePeriodSeconds int    `mapstructure:"grace_period_seconds" validate:"omitempty"`
+}
+
+func Action(action *rules.Action, event *events.Event) (utils.LogLine, error) {
+	podName := event.GetPodName()
+	namespace := event.GetNamespaceName()
+	objects := map[string]string{}
+
+	parameters := action.GetParameters()
+	gracePeriodSeconds := new(int64)
+	if val, ok := parameters["grace_period_seconds"].(int); ok {
+		*gracePeriodSeconds = int64(val)
+	}
+
+	client := kubernetes.GetClient()
+	pod, err := client.GetPod(podName, namespace)
+	if err != nil {
+		return utils.LogLine{
+			Objects: objects,
+			Error:   err.Error(),
+			Status:  "failure",
+		}, err
+	}
+
+	node, err := client.GetNodeFromPod(pod)
+	if err != nil {
+		return utils.LogLine{
+			Objects: objects,
+			Error:   err.Error(),
+			Status:  "failure",
+		}, err
+	}
+
+	objects["node"] = node.Name
+
+	pods, err := client.Clientset.CoreV1().Pods("").List(context.Background(), metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("spec.nodeName=%s", node.GetName()),
+	})
+	if err != nil {
+		return utils.LogLine{
+			Objects: objects,
+			Error:   err.Error(),
+			Status:  "failure",
+		}, err
+	}
+
+	var aggregatedLog utils.LogLine
+	aggregatedLog.Objects = make(map[string]string)
+	var allErrors []string
+	var ignoredPods []string
+
+	for _, p := range pods.Items {
+		line, err, ignored := helpers.VerifyIfPodWillBeIgnored(parameters, client, p, objects)
+		if ignored {
+			// Append ignored pods info
+			aggregatedLog.Output += line.Output + "; "
+			ignoredPods = append(ignoredPods, p.Name)
+			continue
+		}
+		if err != nil {
+			allErrors = append(allErrors, err.Error())
+			continue
+		}
+
+		// Eviction logic if the pod is not ignored
+		err = performEviction(client, p, gracePeriodSeconds)
+		if err != nil {
+			allErrors = append(allErrors, err.Error())
+		}
+	}
+
+	if len(allErrors) > 0 {
+		aggregatedLog.Error = strings.Join(allErrors, "; ")
+		aggregatedLog.Status = "failure"
+	} else {
+		aggregatedLog.Status = "success"
+		aggregatedLog.Output = fmt.Sprintf("Node '%v' drained, ignored pods: %v", node.Name, strings.Join(ignoredPods, ", "))
```

Comment:
Not sure it's useful to list the not-drained pods, for some nodes, it can be hundred of pods. Better to just specify the daemonsets, statefulsets, .. are ignored.

Reply:
ignoredPods[] will only contain pods ignored from the function VerifyIfPod...
, which only ignores pods from either statefulsets, daemonsets or replicasets:
```
				line, err, ignored := helpers.VerifyIfPodWillBeIgnored(parameters, client, p, objects)
		if ignored {
			// Append ignored pods info
			aggregatedLog.Output += line.Output + "; "
			ignoredPods = append(ignoredPods, p.Name)
			continue
		}
```

Reply:
I agree, but the list can be long, and I'm not sure it's useful to have the list in the log line.